File: npm/src/serve/api_chain.js
'use strict'

const fs = require('fs')
const path = require('path')
const { resolveAgentNamespaces, extractRefPath } = require('../validator')
const { normalizeRefSeparator } = require('../lib/pathfmt')

const ROADMAP_STATES = ['wip', 'backlog', 'blocked', 'done', 'abandoned']

/**
 * parseFrontmatter extrai todos os campos de um bloco frontmatter YAML (entre --- e ---).
 * Retorna objeto com chaves/valores string.
 * @param {string} content
 * @returns {Object<string, string>}
 */
function parseFrontmatter(content) {
  const fields = {}
  const lines = content.split('\n')
  let started = false
  let inFm = false
  for (const line of lines) {
    const trimmed = line.trim()
    if (trimmed === '---') {
      if (!started) { started = true; inFm = true; continue }
      break // fecha frontmatter
    }
    if (!inFm) break
    const colonIdx = trimmed.indexOf(':')
    if (colonIdx < 0) continue
    const key = trimmed.slice(0, colonIdx).trim()
    let val = trimmed.slice(colonIdx + 1).trim()
    val = val.replace(/^["']|["']$/g, '')
    if (key) fields[key] = val
  }
  return fields
}

/**
 * extractTitle retorna a primeira linha '# ...' ou o nome do arquivo sem extensão.
 * @param {string} content
 * @param {string} filename
 * @returns {string}
 */
function extractTitle(content, filename) {
  for (const line of content.split('\n')) {
    const t = line.trim()
    if (t.startsWith('# ')) return t.slice(2).trim()
  }
  return filename.replace(/\.md$/, '')
}

/**
 * collectNodes lê arquivos .md de um diretório e retorna nós do grafo.
 * @param {string} dir
 * @param {'adr'|'req'|'roadmap'} type
 * @param {string} state
 * @returns {Array<{id: string, type: string, title: string, state: string, fm: Object}>}
 */
function collectNodes(dir, type, state) {
  const nodes = []
  let files = []
  try {
    files = fs.readdirSync(dir).filter(f => {
      if (!f.endsWith('.md')) return false
      try { return !fs.statSync(path.join(dir, f)).isDirectory() } catch (_) { return false }
    })
  } catch (_) {
    return nodes
  }
  for (const file of files) {
    let content = ''
    try { content = fs.readFileSync(path.join(dir, file), 'utf8') } catch (_) {}
    const fm = parseFrontmatter(content)
    const title = extractTitle(content, file)
    // node ID do grafo — IDENTIFICADOR emitido em JSON (ADR-2026-09-04, D1 categoria 2),
    // e tambem o alvo de toda aresta: resolveRef() abaixo so devolve valores vindos de
    // fileIndex/titleIndex, isto e, sempre um node id — nunca o valor cru do frontmatter.
    // Normalizar o id aqui, portanto, normaliza `edge.to` por construcao.
    //
    // 🔴 O path.join da linha 71 (fs.readFileSync) e uma EXPRESSAO INDEPENDENTE sobre os
    // mesmos operandos: normalizar esta nao pode afeta-la. O caminho de travessia
    // permanece nativo (ADR D2). Espelha internal/serve/api_chain.go:111.
    const id = normalizeRefSeparator(path.join(dir, file))
    nodes.push({ id, type, title, state, fm, content })
  }
  return nodes
}

/**
 * handleChain responde ao GET /api/chain com nodes e edges em JSON.
 * @param {object} cfg
 * @param {http.IncomingMessage} req
 * @param {http.ServerResponse} res
 */
function handleChain(cfg, req, res) {
  const adrDirs = cfg.adrDirs || ['docs/adr']
  const reqDir = cfg.reqDir || 'docs/req'
  const roadmapDir = cfg.roadmapDir || 'docs/roadmaps'
  const namespacing = cfg.roadmapNamespacing || 'flat'

  const allNodes = []

  // ADRs — scan recursivo de subpastas (done/, wip/, etc.)
  for (const adrDir of adrDirs) {
    const subfolders = ['done', 'wip', 'backlog', 'blocked', 'abandoned', '']
    for (const sub of subfolders) {
      const dir = sub ? path.join(adrDir, sub) : adrDir
      const nodes = collectNodes(dir, 'adr', sub || 'unknown')
      allNodes.push(...nodes)
    }
  }

  // REQs
  if (namespacing === 'by_agent') {
    const agents = resolveAgentNamespaces(cfg, reqDir)
    for (const agent of agents) {
      for (const state of ROADMAP_STATES) {
        const dir = path.join(reqDir, agent, state)
        allNodes.push(...collectNodes(dir, 'req', state))
      }
    }
  } else {
    allNodes.push(...collectNodes(reqDir, 'req', 'unknown'))
  }

  // Roadmaps
  if (namespacing === 'by_agent') {
    const agents = resolveAgentNamespaces(cfg, roadmapDir)
    for (const agent of agents) {
      for (const state of ROADMAP_STATES) {
        const dir = path.join(roadmapDir, agent, state)
        allNodes.push(...collectNodes(dir, 'roadmap', state))
      }
    }
  } else {
    for (const state of ROADMAP_STATES) {
      const dir = path.join(roadmapDir, state)
      allNodes.push(...collectNodes(dir, 'roadmap', state))
    }
  }

  // Deduplicar por id (path)
  const nodeMap = new Map()
  for (const n of allNodes) {
    if (!nodeMap.has(n.id)) nodeMap.set(n.id, n)
  }

  // Construir índice por título e nome de arquivo para resolver referências
  const titleIndex = new Map()  // titulo normalizado -> id
  const fileIndex = new Map()   // basename -> id
  for (const [id, n] of nodeMap) {
    const norm = n.title.toLowerCase().trim()
    if (!titleIndex.has(norm)) titleIndex.set(norm, id)
    const base = path.basename(id).replace(/\.md$/, '').toLowerCase()
    if (!fileIndex.has(base)) fileIndex.set(base, id)
  }

  // Construir edges a partir dos campos de frontmatter
  const edges = []
  const edgeSet = new Set()

  function addEdge(from, to) {
    const key = `${from}→${to}`
    if (!edgeSet.has(key)) {
      edgeSet.add(key)
      edges.push({ from, to })
    }
  }

  function resolveRef(val) {
    if (!val) return null
    // Tenta pelo nome de arquivo — path.basename(val), não val cru: o formato canônico
    // gravado por `trackfw req new`/`roadmap new` é o CAMINHO COMPLETO (ADR-2026-08-01), não o
    // basename isolado. Sem o basename() aqui, um vínculo real como
    // "docs/roadmaps/wip/ROADMAP-x.md" nunca batia contra a chave "roadmap-x" do fileIndex —
    // achado do ML-3D: nenhuma aresta REQ/ADR/Roadmap gerada pelo CLI jamais resolvia por este
    // caminho, mascarado porque os testes existentes só cobrem valor já-basename.
    //
    // Por que este arquivo NÃO importa validator.resolveRoadmapRef (ao contrário de
    // internal/serve/api_chain.go, ML-3D): fileIndex é indexado por basename ACROSS todos os
    // estados (wip/backlog/blocked/done/abandoned) desde antes desta correção — um vínculo
    // `roadmap:` "wip/X.md" já resolve contra o node real em "done/X.md" sem fallback extra.
    // O mecanismo do ML-3B/3D (caminho literal gravado com pasta de estado velha) não afeta
    // este CLI porque ele nunca comparou por igualdade de caminho completo — só por basename/
    // título. Ver docs/cli-parity.md, seção "serve: /api/chain".
    const base = path.basename(val).replace(/\.md$/, '').toLowerCase().trim()
    if (fileIndex.has(base)) return fileIndex.get(base)
    // Tenta pelo título
    const norm = val.toLowerCase().trim()
    if (titleIndex.has(norm)) return titleIndex.get(norm)
    return null
  }

  // Campos de link padrão
  const reqFields = (cfg.linkFields && cfg.linkFields.req) || ['REQ:', 'req:']
  const adrFields = (cfg.linkFields && cfg.linkFields.adr) || ['ADR:', 'adr:']
  const roadmapFields = (cfg.linkFields && cfg.linkFields.roadmap) || ['Roadmap:', 'roadmap:']

  for (const [id, node] of nodeMap) {
    const fm = node.fm || {}

    // Checar campos de frontmatter para criar arestas
    for (const [fmKey, fmVal] of Object.entries(fm)) {
      const keyLower = fmKey.toLowerCase()
      if (keyLower === 'req' || reqFields.some(f => f.toLowerCase().replace(':', '') === keyLower)) {
        const target = resolveRef(fmVal)
        if (target && target !== id) addEdge(id, target)
      }
      if (keyLower === 'adr' || adrFields.some(f => f.toLowerCase().replace(':', '') === keyLower)) {
        const target = resolveRef(fmVal)
        if (target && target !== id) addEdge(id, target)
      }
      if (keyLower === 'roadmap' || roadmapFields.some(f => f.toLowerCase().replace(':', '') === keyLower)) {
        const target = resolveRef(fmVal)
        if (target && target !== id) addEdge(id, target)
      }
    }

    // ML-3D: `trackfw req new` (npm/src/generators/req.js) grava `adr: ""` e `roadmap: ""`
    // SEMPRE vazios no frontmatter — o valor real vive no corpo, em "## Linked ADR / ADR:
    // <path>" e "## Linked Roadmap / Roadmap: <path>". O laço de frontmatter acima nunca
    // encontra esse vínculo para uma REQ gerada pelo próprio CLI (não é caso de borda: é o
    // formato canônico inteiro). validator.extractRefPath varre o CONTEÚDO inteiro
    // linha-a-linha (mesma função usada por validateRefTargetsExist) — ponto único de
    // extração, não duplicado aqui.
    for (const field of ['req', 'adr', 'roadmap']) {
      const val = extractRefPath(node.content || '', field)
      if (!val) continue
      const target = resolveRef(val)
      if (target && target !== id) addEdge(id, target)
    }
  }

  // Montar nodes de saída sem o campo fm interno
  const outputNodes = Array.from(nodeMap.values()).map(n => ({
    id: n.id,
    type: n.type,
    title: n.title,
    state: n.state,
  }))

  res.writeHead(200, { 'Content-Type': 'application/json' })
  res.end(JSON.stringify({ nodes: outputNodes, edges }))
}

module.exports = { handleChain }

File: pypi/trackfw/serve/api_chain.py
"""
serve/api_chain.py — Chain API: retorna grafo ADR -> REQ -> ROADMAP.
Espelho Python de internal/serve/api_chain.go e npm/src/serve/api_chain.js.
"""

import os
import re

from trackfw import config as _config
from trackfw.pathfmt import normalize_ref_separator
from trackfw.validator import extract_ref_path

STATES = ["wip", "backlog", "blocked", "done", "abandoned"]


def _extract_frontmatter(content):
    """
    Extrai campos do bloco frontmatter (entre --- e ---).
    Retorna dict com os campos encontrados.
    Suporta valores simples e listas (- item).
    """
    fields = {}
    lines = content.split("\n")
    if not lines or lines[0].strip() != "---":
        return fields

    in_frontmatter = False
    current_key = None
    current_list = []

    for i, raw_line in enumerate(lines):
        line = raw_line.strip()
        if i == 0:
            in_frontmatter = True
            continue
        if line == "---":
            # Flush lista pendente
            if current_key and current_list:
                fields[current_key] = current_list
            break
        if not in_frontmatter:
            break

        if line.startswith("- "):
            # item de lista
            if current_key:
                current_list.append(line[2:].strip().strip("\"'"))
            continue

        colon_idx = line.find(":")
        if colon_idx > 0:
            # Flush lista anterior
            if current_key and current_list:
                fields[current_key] = current_list

            key = line[:colon_idx].strip()
            val = line[colon_idx + 1:].strip().strip("\"'")
            current_key = key
            current_list = []
            if val:
                fields[key] = val
                current_key = None  # valor inline, não lista

    return fields


def _extract_title(content, filename):
    """Extrai título da primeira linha '# ...' ou usa o nome do arquivo."""
    for line in content.split("\n"):
        stripped = line.strip()
        if stripped.startswith("# "):
            return stripped[2:].strip()
    return os.path.splitext(filename)[0]


def _scan_dir(dir_path, node_type, state):
    """Varre um diretório e retorna lista de nodes com seus campos frontmatter."""
    nodes = []
    if not os.path.isdir(dir_path):
        return nodes
    try:
        files = sorted(
            f for f in os.listdir(dir_path)
            if f.endswith(".md") and not os.path.isdir(os.path.join(dir_path, f))
        )
    except OSError:
        return nodes

    for filename in files:
        full_path = os.path.join(dir_path, filename)
        content = ""
        try:
            with open(full_path, "r", encoding="utf-8") as f:
                content = f.read()
        except OSError:
            pass

        fm = _extract_frontmatter(content)
        title = _extract_title(content, filename)
        # node ID do grafo — IDENTIFICADOR emitido em JSON (ADR-2026-09-04, D1
        # categoria 2). Antes do ML-2A a normalização era um .replace() inline aqui;
        # passou a chamar o ponto único do runtime (D3: "não espalhar ReplaceAll pelos
        # chamadores"). Comportamento idêntico, uma noção de formato a menos.
        #
        # 🔴 full_path (acima) permanece nativo e é o que vai a open() — a normalização
        # é de saída, não de travessia (ADR D2).
        rel_path = normalize_ref_separator(os.path.relpath(full_path, os.getcwd()))

        node = {
            "id": rel_path,
            "type": node_type,
            "title": title,
            "state": state,
            "frontmatter": fm,
            "content": content,
        }
        nodes.append(node)

    return nodes


def get_chain(cfg):
    """
    Retorna grafo { nodes: [...], edges: [...] } representando
    a cadeia ADR → REQ → ROADMAP.
    """
    adr_dirs = cfg.get("adr_dirs", ["docs/adr"])
    req_dir = cfg.get("req_dir", "docs/req")
    roadmap_dir = cfg.get("roadmap_dir", "docs/roadmaps")
    namespacing = cfg.get("roadmap_namespacing", "flat")

    nodes = []
    edges = []

    # --- ADRs ---
    for adr_dir in adr_dirs:
        # Suporte a by_agent: verificar subpastas done/wip/...
        if namespacing == "by_agent":
            agents = _config.resolve_agent_namespaces(cfg, adr_dir)
            for agent in agents:
                for state in STATES:
                    nodes.extend(_scan_dir(os.path.join(adr_dir, agent, state), "adr", state))
        else:
            # flat: pode haver subpastas done/wip ou arquivos direto
            has_state_dirs = any(
                os.path.isdir(os.path.join(adr_dir, s)) for s in STATES
            )
            if has_state_dirs:
                for state in STATES:
                    nodes.extend(_scan_dir(os.path.join(adr_dir, state), "adr", state))
            else:
                nodes.extend(_scan_dir(adr_dir, "adr", "done"))

    # --- REQs ---
    if namespacing == "by_agent":
        agents = _config.resolve_agent_namespaces(cfg, req_dir)
        for agent in agents:
            for state in STATES:
                nodes.extend(_scan_dir(os.path.join(req_dir, agent, state), "req", state))
    else:
        has_state_dirs = any(os.path.isdir(os.path.join(req_dir, s)) for s in STATES)
        if has_state_dirs:
            for state in STATES:
                nodes.extend(_scan_dir(os.path.join(req_dir, state), "req", state))
        else:
            nodes.extend(_scan_dir(req_dir, "req", "unknown"))

    # --- Roadmaps ---
    if namespacing == "by_agent":
        agents = _config.resolve_agent_namespaces(cfg, roadmap_dir)
        for agent in agents:
            for state in STATES:
                nodes.extend(_scan_dir(os.path.join(roadmap_dir, agent, state), "roadmap", state))
    else:
        for state in STATES:
            nodes.extend(_scan_dir(os.path.join(roadmap_dir, state), "roadmap", state))

    # --- Construir índice id → node ---
    by_id = {n["id"]: n for n in nodes}

    # --- Construir índice basename → node (para match por nome de arquivo) ---
    by_basename = {}
    for n in nodes:
        basename = os.path.basename(n["id"])
        by_basename.setdefault(basename, []).append(n)

    def _find_node_by_ref(ref):
        """Tenta encontrar node pelo id exato ou pelo basename.

        ML-3D: basename(ref), não ref cru — o formato canônico gravado por
        `trackfw req new`/`roadmap new` (ADR-2026-08-01) é o CAMINHO COMPLETO
        (ex.: "docs/roadmaps/wip/ROADMAP-x.md"), não um basename isolado. Sem
        aplicar basename() aqui, um vínculo real nunca batia contra as chaves
        de by_basename (que são sempre basenames) — a mesma classe de defeito
        corrigida em npm/src/serve/api_chain.js:resolveRef.
        """
        ref = ref.strip()
        if ref in by_id:
            return by_id[ref]
        base = os.path.basename(ref)
        candidates = by_basename.get(base, []) or by_basename.get(base + ".md", [])
        if candidates:
            return candidates[0]
        return None

    edge_seen = set()

    def _add_edge(from_id, to_id):
        if from_id == to_id:
            return
        key = (from_id, to_id)
        if key in edge_seen:
            return
        edge_seen.add(key)
        edges.append({"from": from_id, "to": to_id})

    def _resolve_field_edges(node, field, fm_ref):
        """Resolve um campo de vínculo (req/adr/roadmap) para o node, tentando o
        frontmatter primeiro (fm_ref) e, se vazio, o valor extraído do CORPO do
        arquivo via extract_ref_path.

        ML-3D — achado: `trackfw req new` (trackfw/generators/req.py) grava
        `adr: ""` e `roadmap: ""` SEMPRE vazios no frontmatter; o valor real
        vive em "## Linked ADR / ADR: <path>" e "## Linked Roadmap /
        Roadmap: <path>", no corpo. fm_ref sozinho NUNCA resolvia o vínculo
        de uma REQ gerada pelo próprio CLI — não era caso de borda, era o
        formato canônico inteiro nunca resolvendo.
        """
        refs = []
        if isinstance(fm_ref, str) and fm_ref:
            refs.append(fm_ref)
        elif isinstance(fm_ref, list):
            refs.extend(r for r in fm_ref if r)
        if not refs:
            body_ref = extract_ref_path(node.get("content", ""), field)
            if body_ref:
                refs.append(body_ref)
        for ref in refs:
            target = _find_node_by_ref(ref)
            if target:
                _add_edge(node["id"], target["id"])

    # --- Construir arestas ---
    for node in nodes:
        fm = node.get("frontmatter", {})

        # REQ → ADR
        if node["type"] == "req":
            _resolve_field_edges(node, "ADR", fm.get("adr", ""))
            # REQ → ROADMAP (achado do ML-3D: faltava por completo — nem o campo de
            # frontmatter nem o do corpo eram lidos para REQ→Roadmap antes desta correção)
            _resolve_field_edges(node, "Roadmap", fm.get("roadmap", ""))

        # ROADMAP → REQ / ROADMAP → ADR
        if node["type"] == "roadmap":
            _resolve_field_edges(node, "REQ", fm.get("req", ""))
            _resolve_field_edges(node, "ADR", fm.get("adr", ""))

    # Remover frontmatter e conteúdo do output (não deve ir para o cliente)
    output_nodes = [
        {k: v for k, v in n.items() if k not in ("frontmatter", "content")}
        for n in nodes
    ]

    return {"nodes": output_nodes, "edges": edges}
